lib/utils/task_queue/worker_queue.dart

```
@@ -0,0 +1,85 @@
+import 'dart:async';
+import 'dart:collection';
+
+import 'package:fluffychat/utils/task_queue/task.dart';
+import 'package:matrix/matrix.dart';
+
+typedef OnTaskCompleted = void Function(String? taskId);
+
+abstract class WorkerQueue {
+  final Queue<Task> _queue = Queue<Task>();
+  Completer? _completer;
+
+  String get workerName;
+
+  Queue<Task> get queue => _queue;
+
+  Task? _runningTask;
+
+  Future addTask(Task task) {
+    _queue.add(task);
+    Logs().i(
+      'WorkerQueue<$workerName>::addTask(): QUEUE_LENGTH: ${_queue.length}',
+    );
+    return _processTask();
+  }
+
+  bool isTaskIdInQueue(String taskId) {
+    return _queue.any((task) => task.id == taskId) ||
+        _runningTask?.id == taskId;
+  }
+
+  Future _processTask() async {
+    if (_completer != null) {
+      return _completer!.future;
+    }
+    Logs().i(
+      'WorkerQueue<$workerName>::addTask(): QUEUE is ready',
+    );
+    _completer = Completer();
+    if (_queue.isNotEmpty) {
+      final firstTask = _queue.removeFirst();
+      Logs().i('WorkerQueue<$workerName>::_processTask(): ${firstTask.id}');
+      _runningTask = firstTask;
+      firstTask
+          .execute()
+          .then(_handleTaskExecuteCompleted)
+          .catchError(_handleTaskExecuteError)
+          .whenComplete(() {
+        firstTask.onTaskCompleted?.call();
+        _runningTask = null;
+      });
+    } else {
+      _completer?.complete();
+    }
+    return _completer!.future;
+  }
+
+  void _handleTaskExecuteCompleted(dynamic value) {
+    Logs().i('WorkerQueue<$workerName>::_handleTaskExecuteCompleted(): $value');
+    _completer?.complete(value);
+    _releaseCompleter();
+    if (_queue.isNotEmpty) {
+      _processTask();
+    }
+  }
+
+  void _handleTaskExecuteError(error) {
+    Logs().i('WorkerQueue<$workerName>::_handleTaskExecuteError(): $error');
+    _completer?.complete(error);
+    _releaseCompleter();
+    if (_queue.isNotEmpty) {
+      _processTask();
+    }
+  }
```

try to focus on error handling too. Try not to swallow error